File: 0015-3sum/0015-3sum.cs
public class Solution {
    public IList<IList<int>> ThreeSum(int[] nums)
    {
        Array.Sort(nums);

        var res = new List<IList<int>>();

        for (int i = 0; i < nums.Length - 2; i++)
        {
            int l = i + 1;
            int r = nums.Length - 1;

            while (l < r)
            {
                int sum = nums[l] + nums[r] + nums[i];

                if (sum == 0)
                {
                    var triplet = new[] { nums[l], nums[i], nums[r] };

                    res.Add(triplet);

                    while (l < i && nums[l] == triplet[0]) l++;
                    while (i < r && nums[r] == triplet[2]) r--;
                }
                else if (sum < 0)
                {
                    l++;
                }
                else
                {
                    r--;
                }
            }

            while (i + 1 < nums.Length - 1 && nums[i + 1] == nums[i])  i++;
        }

        return res;
    }
}

File: 0015-3sum/0015-3sum.go
func threeSum(nums []int) [][]int {
	sort.Slice(nums, func(i, j int) bool {
		return nums[i] < nums[j]
	})

	n := len(nums)

	res := make([][]int, 0)

	for i := 0; i < n - 2; i++ {
        if i != 0 && nums[i] == nums[i - 1] {
            continue
        }
        
		c := nums[i]

		l := i + 1
		r := n - 1

		for l < r {
			a := nums[l]
			b := nums[r]

			sum := a + b + c

			if sum == 0 {
				res = append(res, []int{a, b, c})

				for l < r && nums[l] == a {
					l++
				}

				for l < r && nums[r] == b {
					r--
				}
			} else if sum > 0 {
				r--
			} else {
				l++
			}
		}
	}

	return res
}